### ISSUE TITLE
I use the ebpf mode to obtain the profile data of the c++ program, and found that there are a lot of unknown data in the c++ program, what is going on, and how to solve it

### ISSUE THREAD

**Issue by ly0511** [grafana/pyroscope#1970]

**Comment by AshinZ** [comment]
have you compile c++ with -g? or  you can use `file ` to find whether this file has debug info.

**Comment by ingangi** [comment]
@AshinZ  I have the same issue, my binary was compiled by -g, and below is the output of `file` cmd:
```
ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=xxx, for GNU/Linux 3.2.0, with debug_info, not stripped, too many notes (256)
```

**Comment by korniltsev** [comment]
Can you share the flamegraph? Is your program compiled with omit-frame-poibters?

**Comment by AshinZ** [comment]
@ly0511 @ingangi  maybe you have used -O option when you complile your c++ code. you can re-compile your c++ code with -fno-omit-frame-pointer  option.

**Comment by chintanp** [comment]
@ly0511 @ingangi I am trying to use Pyroscope for C/C++ app as well and not sure how to profile yet. Could you share how you are getting your profiles? My C app is in a container. Do you have a similar setup?

**Comment by Thor-wl** [comment]
> @ly0511 @ingangi I am trying to use Pyroscope for C/C++ app as well and not sure how to profile yet. Could you share how you are getting your profiles? My C app is in a container. Do you have a similar setup?

+1

**Comment by Thor-wl** [comment]
> @ly0511 @ingangi I am trying to use Pyroscope for C/C++ app as well and not sure how to profile yet. Could you share how you are getting your profiles? My C app is in a container. Do you have a similar setup?

+1

**Comment by chintanp** [comment]
> @Thor-wl @chintanp I use ebpf mode to get the c++ profile,you can try this way

Yes, I have been trying the ebpf mode but with no success. Could you share the docker-compose file you use? I am wondering if the C process and Pyroscope need to be in the same container ? And if so, does one of the two (Pyroscope and the C app) have to run in the background?

**Comment by korniltsev** [comment]
The current implementation does only framepointer unwinding.
We have plans for dwarf unwinding.
In the mean time I suggest you can try recompiling your c/c++ app and libs with `-fno-omit-frame-pointer`. Note: compiler will take a gp register for stack framepointer so performance may or may not degrade a little

**Comment by tvesely** [comment]
Is dwarf stack unwinding on the roadmap, or is it up in the air?

**Comment by korniltsev** [comment]
> Is dwarf stack unwinding on the roadmap, or is it up in the air?
up in the air

**Comment by luweglarz** [comment]
Hey guys, came across this issue because I am facing the same problems with unknown symbols when the frame pointers are omitted, but then I saw this [PR](https://github.com/grafana/pyroscope/pull/3799), does this intent the implementation of DWARF stack unwinding? From what I read from this PR the goal is to implement this symbolization at Pyroscope read path so I'm not sure if this is gonna solve the stack unwinding without frame pointer, i thought such thing should be done on agent side, maybe I am missing a piece

**Comment by korniltsev** [comment]
Improved native languages unwinding is coming with https://github.com/grafana/alloy/pull/2920 which is based on a fork of openetelemetry-ebpf-profiler.

[The linked PR](https://github.com/grafana/pyroscope/pull/3799) symbolizes unsymbolized data that we expect to be ingested from opentelemetry-ebpf-profiler / otel profiling collcetor. It is not related to unwinding itself.

**Comment by luweglarz** [comment]
Awesome! excited to test it !

**Comment by ruslaniv** [comment]
It still not clear if Pyroscope uses anything besides frame pointers to unwind the stack? 
Has anything like DWARF or .eh_frame been implemented? I can't find any information about this (

**Comment by luweglarz** [comment]
@ruslaniv  Alloy with pyroscope.ebpf use the https://github.com/open-telemetry/opentelemetry-ebpf-profiler as underlying profiler, which supports advanced unwinding and symbolization without frame pointer
